_posts/2021-05-28-why-avoid-an-orm.md

```diff
@@ -0,0 +1,231 @@
+ 
+The most obvious (and usually the best) configuration is to configure all relationships as lazily fetched, and everything else (that is, "ordinary data") as eagerly fetched (oddly, [Java Persistence API](https://en.wikibooks.org/wiki/Java_Persistence) decided to default to fetching to-one relationships eagerly). In this case on `line 2` a single query is made for all departments, and on `line 3` one query is made for each department to get its employees. On `line 4` no more queries would be made since employee names would be fetched already on `line 3`. So if there are 10 departments, this is 10+1 queries. If there are 100 departments, this is 100+1 queries.
+ 
+This is going to blow up if there are more departments. You can say ”hey, that code is easy to fix” but the problem is that in real life that object graph traversing is separated into multiple different functions each performing their own thing, and they cannot know when they are doing database fetches.
+ 
+Imagine if you only wanted to print the number of employees for each department. You’d be performing the same amount of queries and still fetching half the database to memory for nothing. Some ORMs optimize for this by providing collection _proxies_, which they only populate with item references (proxies) without actual data. But it’s only a hack at best, and we get to step into the rabbit hole of proxies.
+ 
+ORMs of course provide ways to avoid n+1 problems. One way is to configure the fetching strategy for a relationship to be eager so that you can say that the employees of a department are always fetched from the database whenever the department data is fetched. But imagine the performance when there are departments with thousands of employees and your use case is only interested in the department data and not the employees.
+ 
+This can be improved by providing different fetching strategies for different use cases, but since the mapping configuration is more or less static, you can’t handle different needs easily. Well, you can (for example by using annotations/metadata to specify different fetching strategies that can be enabled one way or another), but you need to fall back to using some kind of query language where you can specify how much data you want to fetch. And since you are falling back to a query language, why not just use a query language in the first place and forget about object graphs.
+ 
+Another problem is that maybe you have pagination and you want to display only 50 employees at a time. But you can’t specify this in an object graph since all you are saying is `getEmployees()` and doing the pagination afterward. ORMs handle this by not fetching all employees in one query, but some configurable amount at a time. Again, the amount of actual queries (or physical database fetches of the result of a single query) depends on static configuration and you’d need to fall back to a query language to vary it for different use cases.
```

Comment:
Perhaps it's just me not having come across this case, but I have never seen N + 1 issues with pagination queries. I'd also guess that most ORMs also provide pagination tools at both entity manager and query builder level.

I've definitely seen pagination issues with more complex ORM queries though. And of course pagination is a whole can of worms in general. :smile:

Reply:
Hah, you are right, I'm not really talking about just "n+1 queries", I'm talking about the amount of queries in general!
I changed the heading and made some small changes, how does it feel now?

Reply:
Yup, fits the heading better.